src/graph_stats.cc

```diff
@@ -0,0 +1,160 @@
+#include "graph_stats.h"
+#include "state.h"
+#include "graph.h"
+
+// Standard headers
+#include <limits>
+#include <set>
+#include <queue>
+#include <vector>
+#include <cassert>
+
+namespace {
+
+class GetGraphHeight
+{
+public:
+  GetGraphHeight()
+    : _heights()
+  {}
+
+private:
+  typedef map<Node*, size_t> Heights;
+  Heights _heights;
+
+public:
+  void Reset() {
+    _heights.clear();
+  }
+
+  size_t operator()(const vector<Node*>& root_nodes) {
+    Reset();
+    return GetMaxHeight(root_nodes);
+  }
+
+private:
+  size_t operator()(Node* node) {
+    size_t& height = _heights[node];
+    if (height == 0) {
+      if (Edge* e = node->in_edge()) {
+        height = GetMaxHeight(e->inputs_) + 1;
+      } else {
+        height = 1;
+      }
+    }
+    return height;
+  }
+
+  size_t GetMaxHeight(const vector<Node*>& nodes) {
+    size_t height = numeric_limits<size_t>::min();
+    for (vector<Node*>::const_iterator i = nodes.begin();
+         i != nodes.end(); ++i) {
+      height = max(height, (*this)(*i));
+    }
+    return height;
```

Comment:
Isn't this loop equivalent to something like

    return std::max_element(nodes.begin(), nodes.end(), LesserHeightOf);